### ISSUE TITLE
Sample exams template research

### ISSUE THREAD

**Issue by danopolan** [danopolan/istqb_latex#68]
Elaborate on possibilities to tweak current syllabus template in order to support generation of Sample Exam documents. Namely:

- Sample Questions (https://istqb-main-web-prod.s3.amazonaws.com/media/documents/ISTQB_CTFL_Sample-Exam-Questions_v4.0.pdf)
- Sample Answers (https://istqb-main-web-prod.s3.amazonaws.com/media/documents/ISTQB_CTFL_Sample-Exam-Answers_v4.0.pdf)
- Exam Structure and Rules (https://istqb-main-web-prod.s3.amazonaws.com/media/documents/ISTQB_Exam-Structure-Tables_v1_6.pdf)

Beware, we are not trying to generate documents which are identical to the existing ones. We should research possible formatting of questions (stem, question, answers options, justification, answer key) in Markdown to generate documents which has the same core as syllabus just a different content.

Possibly we could generate Exam document (stem, question, answers options), Exam with Answers document (stem, question, answers options, justification, answer key).

Exam Structure and Rules document should be split by chapters, but I am not sure if it makes sense to convert such table into Markdown.

**Comment by Witiko** [comment]
For the sample questions and answers, we can have a single YAML source that would contain all questions and answers. For example, here is how sample questions & answers 1, 5, and 6 might look:

``` yaml
1:
  learning-objective: FL-1.1.1
  k-level: K1
  number-of-points: 1
  question: Which of the following statements describe a valid test objective?
  answers:
    a: To prove that there are no unfixed defects in the system under test
    b: To prove that there will be no failures after the implementation of the system into production
    c: To reduce the risk level of the test object and to build confidence in the quality level
    d: To verify that there are no untested combinations of inputs
  correct: c
  explanation: |
    a) Is not correct. It is impossible to prove that there are no defects anymore in the system under test.
       See testing principle 1
    b) Is not correct. See testing principle 7
    c) Is correct. Testing finds defects and failures which reduces the level of risk and at the same time
       gives more confidence in the quality level of the test object
    d) Is not correct. It is impossible to test all combinations of inputs (see testing principle 2)
5:
  learning-objective: FL-1.4.2
  k-level: K2
  number-of-points: 1
  question: |
    Which of the following factors (i-v) have SIGNIFICANT influence on the test process?

    i. The SDLC
    ii. The number of defects detected in previous projects
    iii. The identified product risks
    iv. New regulatory requirements forcing
    v. The number of certified testers in the organization
  answers:
    a: i, ii have significant influence; iii, iv, v have not
    b: i, iii, iv have significant influence; ii, v have not
    c: ii, iv, v have significant influence; i, iii have not
    d: iii, v have significant influence; i, ii, iv have not
  correct: b
  explanation: |
    i.   Is true. The SDLC has an influence on the test process
    ii.  Is false. The number of defects detected in previous projects may have
         some influence, but this is not as significant as i, iii and iv
    iii. Is true. The identified product risks are one of the most important
         factors influencing the test process
    iv.  Is true. Regulatory requirements are important factors influencing the 
         test process
    v.   Is false. The test environment should be a copy of the production
         environment but has no significant influence on the test process

    Hence b is correct.
6:
  learning-objective: FL-1.4.5
  k-level: K2
  number-of-points: 1
  question: Which TWO of the following tasks belong MAINLY to a testing role?
  answers:
    a: Configure test environments
    b: Maintain the product backlog
    c: Design solutions to new requirements
    d: Create the test plan
    e: Report on achieved coverage
  correct: [a, e]
  explanation: |
    a) Is correct. This is done by the testers
    b) Is not correct. The product backlog is built and maintained by the product owner
    c) Is not correct. This is done by the development team
    d) Is not correct. This is a managerial role
    e) Is correct. This is done by the testers
```

**Comment by Witiko** [comment]
For the exam structure and rules, I assume each row in the table on page 3 corresponds to a single syllabus and a single set of questions. Therefore, if we had a YAML document for every syllabus, we could generate the table automatically. I am not sure about the remaining tables in the document: is there an opportunity for automatic generation there?

**Comment by danopolan** [comment]
This looks great. Only a few additional questions from my side:
1. Would it be possible to incorporate images and tables into the YAML file, respectively to the questions?
2. What will be the effort to implement the conversion from YAML to PDF?
3. Can we add more parameters at the beggining? Like:
```
duration: 60/75 min
questions: 40
max-score: 40
pass-score: 65%
```
4. Exam Structure & Rules is based on Learning Objectives. All Learning Objectives from a syllabus should be examinable (you can make an exam question out of it), but the specific structure of the exam is defined by Exam Structure & Rules. Based on this document we create an exam set of 40 questions which is given to the candidate.

**Comment by Witiko** [comment]
> 1. Would it be possible to incorporate images and tables into the YAML file, respectively to the questions?

Yes, the questions and the explanation can contain arbitrary markdown text.

> 2. What will be the effort to implement the conversion from YAML to PDF?

For Sample Questions, we can simply reuse the current template for syllabi. The expected effort for the conversion from YAML to PDF is 2 to 4 hours.

For Sample Answers, we will need to add support for a landscape layout. The expected effort for the conversion from YAML to PDF is 4 to 6 hours.

> 3. Can we add more parameters at the beggining? Like:

Not a problem, although it seems best to keep things structured and numeric.

``` yaml
duration: [60, 75]
questions: 40
max-score: 40
pass-score: 65
```

This will allow us to use the YAML data for computation, if necessary, not just for typesetting.

> 4. Exam Structure & Rules is based on Learning Objectives. All Learning Objectives from a syllabus should be examinable (you can make an exam question out of it), but the specific structure of the exam is defined by Exam Structure & Rules. Based on this document we create an exam set of 40 questions which is given to the candidate.

I see. The way I see it, only the table on page 3 can be automatically generated from question YAML data. The remainder of the document are mostly manually-written tables with complex layout, which seem best prepared in LaTeX or a different tool altogether.

**Comment by danopolan** [comment]
Great, so if you can, please create tasks for Sample Exam and Answers generation.
Otherwise I accept 1. and 3. points.
Point 4 will be dropped for now.

**Comment by Witiko** [comment]
> Great, so if you can, please create tasks for Sample Exam and Answers generation.

I am not sure what "creating tasks" means. Would you like me to open issues for Sample Exam and Sample Answers templates and start working on them?

**Comment by danopolan** [comment]
I meant creating Issues :) But I can do it as well, if you want to be sure, you have the info you need in the Issue from my side.

**Comment by Witiko** [comment]
@danopolan I opened a new issue for the generation of Sample Exam Questions and Answers documents in https://github.com/danopolan/istqb_latex/issues/85. If you agree with the acceptance criteria, then we can close this issue.

**Comment by Witiko** [comment]
Closed via #87.
